### PR TITLE
Support Tensor alias annotations for native_functions.yaml

### PR DESCRIPTION
**Pull request by cpuhrsch**
Adds Tensor alias annotations. 

This isn't a full implementation of alias annotations, but that isn't required to increase compliance with the JIT signature schema. There are some sanity checks within native_parse.py for their usage, which can also help overall correctness. Otherwise, this exists solely for further alignment between the JIT signature schema and the native_functions.yaml func schema.

This gets us to ~85% matches.

### CODE REVIEW

**Comment by cpuhrsch** [comment]
This is missing updates to comments and README, but the code is good to review.

File: aten/src/ATen/native/native_functions.yaml
Comment (ezyang) [inline aten/src/ATen/native/native_functions.yaml]: So, is it true that alias annotation serves two purposes: (1) indicate aliasing has occurred and (2) indicate that the argument in question will mutate?

File: aten/src/ATen/native_parse.py
Comment (ezyang) [inline aten/src/ATen/native_parse.py]: LOL. I hope we don't need this!

File: aten/src/ATen/native/native_functions.yaml
Comment (cpuhrsch) [reply aten/src/ATen/native/native_functions.yaml]: Yes and a lot more when complete.

The full description is [here](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/operator.cpp#L101).

File: aten/src/ATen/native_parse.py
Comment (cpuhrsch) [reply aten/src/ATen/native_parse.py]: Sure, I can change that

File: aten/src/ATen/native/native_functions.yaml
Comment (ezyang) [reply aten/src/ATen/native/native_functions.yaml]: OK, looking forward to when you add this to the README :)

File: aten/src/ATen/native_parse.py
Comment (cpuhrsch) [reply aten/src/ATen/native_parse.py]: Here is the parsing code: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/operator.cpp#L107

I can update the regex to be more precise, but there isn't anything that requires that precision right now.

Comment (ezyang) [reply aten/src/ATen/native_parse.py]: OK, up to you. This is mostly a "future-proofing" consideration.

File: aten/src/ATen/native_parse.py
Comment (ezyang) [inline aten/src/ATen/native_parse.py]: Isn't this violated by `_copy_same_type_`? I'm curious why this isn't triggering. https://github.com/pytorch/pytorch/blob/1c0504b889d256391a5a51a3ca8dd07f41fd649d/aten/src/ATen/native/native_functions.yaml#L559

File: aten/src/ATen/native_parse.py
Comment (ezyang) [inline aten/src/ATen/native_parse.py]: nit: transition

**Review by ezyang — APPROVED** [review]
Looks legit

File: aten/src/ATen/native_parse.py
Comment (gchanan) [reply aten/src/ATen/native_parse.py]: This function is basically trying to do everything.

IMO the way you want to structure this is so you can just rip out the legacy translations, but there's all sorts of things in here, like legacy sanitation (why?), default handling, size parsing, etc.  How about something like this:

1) something that just does try argument legacy translations.  This only needs to look at the type specifier.  It doesn't need to pull out the numbers inside `[]` -- you have to do that in whether legacy translation exists or not, it shouldn't be in here.
2) something that does legacy default translation.  It needs to have the type because the translations are type specific.  But you can have it be based on the pre-translated (i.e. correct) type, so if you remove the specific type translation, nothing has to change here.

Then leave everything else the same, but write the other parts to operate on pre-translated types, so instead of parsing the `3` out of `IntList[3]` you parse it out of `int[3]`.  The translations are just for passing to downstream components, native_parse should be "clean" otherwise.

File: tools/autograd/gen_variable_type.py
Comment (gchanan) [inline tools/autograd/gen_variable_type.py]: shouldn't this already have a name from declarations.yaml?

File: aten/src/ATen/native_parse.py
Comment (ezyang) [reply aten/src/ATen/native_parse.py]: In defense of this PR as is, it does deduplicate two code paths. That seems like a win in my book. So I'd be OK with trying to refactor this code even more as we keep working on this path, while committing this for now.

Comment (cpuhrsch) [reply aten/src/ATen/native_parse.py]: The distinction between legacy translations and non-legacy translations is not clear. Once the func schema complies with the JIT signature schema, we will start ripping out pretty much all the translations. That will happen one by one. Whether that corresponds to a legacy translation or something before then won't matter much. It is however important to make it very easy to understand what and how exactly things are being mapped. Separating this into legacy and non-legacy will make that more difficult to understand.

File: tools/autograd/gen_variable_type.py
Comment (cpuhrsch) [reply tools/autograd/gen_variable_type.py]: This only does a substitution. Using a specific name and passing that through will follow later in conjunction with [this PR](https://github.com/pytorch/pytorch/pull/16253).

File: aten/src/ATen/native_parse.py
Comment (cpuhrsch) [reply aten/src/ATen/native_parse.py]: It's not a keyword Tensor argument, but positional.

**Comment by cpuhrsch** [comment]
@ezyang @gchanan - I updated the README to cover annotations in context of our function schema and I'm still hunting code-level comments that mention result/result0 etc., but aside from that the documentation should be good to go for review. After that I'd want to merge this PR.

File: aten/src/ATen/native/README.md
Comment (gchanan) [inline aten/src/ATen/native/README.md]: there's already a bullet point about Tensor / Tensor? above -- is there a reason this is separate?

File: aten/src/ATen/native/README.md
Comment (gchanan) [inline aten/src/ATen/native/README.md]: I know this is coming from the original JIT docs (so not your issue), but it could use some improvement:
- "is in set" is never defined.
- don't use 'a', it's a word and makes the statement harder to parse.

File: aten/src/ATen/native/README.md
Comment (gchanan) [inline aten/src/ATen/native/README.md]: IMO this should be its own section, instead of linking to the variants section.  It's pretty meaty.

File: aten/src/ATen/native/README.md
Comment (gchanan) [inline aten/src/ATen/native/README.md]: I'm confused how the non-inplace, non-out variant got in here.  Maybe you just want to introduce it above?

File: aten/src/ATen/native/README.md
Comment (gchanan) [inline aten/src/ATen/native/README.md]: 'situations'.

Comment (gchanan) [inline aten/src/ATen/native/README.md]: be clear about what you introducing.  Say something like "Let's revisit the previous native function declarations and see the effects of adding annotations."

File: aten/src/ATen/native/README.md
Comment (gchanan) [inline aten/src/ATen/native/README.md]: `self` instead of self.

Also I feel like the second sentence is missing something?  What does it indicate about the return value?

File: aten/src/ATen/native/README.md
Comment (gchanan) [inline aten/src/ATen/native/README.md]: I'm not sure about this -- that's the name in the python frontend, why wouldn't we need the name to be the same?

File: aten/src/ATen/native/README.md
Comment (gchanan) [inline aten/src/ATen/native/README.md]: `out` instead of out.

Comment (gchanan) [reply aten/src/ATen/native/README.md]: I'm not sure what "if so" means here.

File: aten/src/ATen/native/README.md
Comment (gchanan) [inline aten/src/ATen/native/README.md]: What do you check?

File: aten/src/ATen/native/README.md
Comment (gchanan) [inline aten/src/ATen/native/README.md]: there was a lot of discussion about *may* vs *must* and I'm admittedly still confused...

The situation discussed was `contiguous` which may return the same memory, if the Tensor is already contiguous.

If *may* is the correct answer, then an acceptable annotation is:
`contiguous(Tensor(a) self)-> Tensor(a)`

If *must* is the correct answer, then an acceptable annotation is:
`contiguous(Tensor(a) self)-> Tensor(a|b)` (i.e. it *must* be memory from `a` or memory from newly-introduced `b`).

Of course the second definition is also acceptable for *may*, but you'd never do that...

**Review by gchanan — COMMENTED** [review]
Some comments on the annotation writeup -- thanks for doing this, it's much clearer now.  I don't think this needs to block merging, since the native_functions.yaml changes are merge conflicts waiting to happen...

File: aten/src/ATen/native_parse.py
Comment (gchanan) [reply aten/src/ATen/native_parse.py]: regarding @ezyang's point about deduplicating two code paths, that's true although the code paths should at least be partially separate (return arguments can't have defaults, now they are parsed the same way and we just pretend the default specification wasn't there.

File: aten/src/ATen/native/README.md
Comment (cpuhrsch) [reply aten/src/ATen/native/README.md]: Yeah it's superfluous. It's in essence trying to showcase the power of annotations. I removed it

File: aten/src/ATen/native/README.md
Comment (cpuhrsch) [reply aten/src/ATen/native/README.md]: It is indeed *may* not *must*. We don't use annotations rigorously yet as far as I can tell. This will become more apparent as we move along the unification and is a feature of this process.

File: aten/src/ATen/native/README.md
Comment (gchanan) [inline aten/src/ATen/native/README.md]: nit: I don't think you need the first `a` and the second a should be `a`.

File: aten/src/ATen/native/README.md
Comment (gchanan) [inline aten/src/ATen/native/README.md]: based on the explanation here, I'd expect a more full explanation below, but it's not there.  Maybe just remove this example until we have an actual example annotation?

**Comment by cpuhrsch** [comment]
For references, flake8 lint errors were unrelated to this PR.

**Pull request closed without merging**
